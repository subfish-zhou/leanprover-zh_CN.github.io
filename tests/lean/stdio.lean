
open IO.FS
def usingIO {α} (x : IO α) : IO α := x

#eval usingIO do
let out ← IO.getStdout;
out.putStrLn "print stdout"

#eval usingIO do
let err ← IO.getStderr;
(err.putStr "print stderr" : IO Unit)

open usingIO IO

def test : IO Unit := do
FS.withFile "stdout1.txt" IO.FS.Mode.write $ fun h₁ => do
{ let h₂ ← FS.Handle.mk "stdout2.txt" IO.FS.Mode.write;
  withStdout (Stream.ofHandle h₁) $ do
    println "line 1";
    tryCatch
      ( do
        withStdout (Stream.ofHandle h₂) $ println "line 2";
        throw $ IO.userError "my error" )
      ( fun e => println e );
    println "line 3" };
println "line 4";
println "\n> stdout1.txt";
readFile "stdout1.txt" >>= print;
println "\n> stdout2.txt";
readFile "stdout2.txt" >>= print

#eval test
