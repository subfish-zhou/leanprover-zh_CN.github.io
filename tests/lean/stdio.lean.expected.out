print stdout
print stderr
line 4

> stdout1.txt
line 1
my error
line 3

> stdout2.txt
line 2
